service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java

```diff
@@ -0,0 +1,187 @@
+            .addParameter(JobMapKeys.DESCRIPTION.getKeyName(), description)
+            .addParameter(RunImputationJobFlightMapKeys.PIPELINE_ID, pipeline.getId())
+            .addParameter(
+                RunImputationJobFlightMapKeys.PIPELINE_INPUT_DEFINITIONS,
+                pipeline.getPipelineInputDefinitions())
+            .addParameter(
+                RunImputationJobFlightMapKeys.USER_PROVIDED_PIPELINE_INPUTS, userProvidedInputs)
+            .addParameter(
+                RunImputationJobFlightMapKeys.CONTROL_WORKSPACE_ID,
+                pipeline.getWorkspaceId().toString())
+            .addParameter(
+                RunImputationJobFlightMapKeys.WDL_METHOD_NAME, pipeline.getWdlMethodName())
+            .addParameter(JobMapKeys.RESULT_PATH.getKeyName(), resultPath);
+
+    jobBuilder.submit();
+
+    logger.info("Created {} pipelineRun with jobId {}", pipelineName, jobId);
+
+    return pipelineRun;
+  }
+
+  public PipelineRun writePipelineRunToDb(
+      UUID jobUuid,
+      String userId,
+      Long pipelineId,
+      CommonPipelineRunStatusEnum status,
+      String description,
+      String resultUrl,
+      Map<String, Object> pipelineInputs) {
+
+    // write pipelineRun to database
+    PipelineRun pipelineRun =
+        new PipelineRun(jobUuid, userId, pipelineId, status.toString(), description, resultUrl);
+    PipelineRun createdPipelineRun = writePipelineRunToDbThrowsDuplicateException(pipelineRun);
+
+    String pipelineInputsAsString;
+    try {
+      // do this to write the pipeline inputs without writing the class name
+      pipelineInputsAsString = objectMapper.writeValueAsString(pipelineInputs);
+    } catch (JsonProcessingException e) {
+      throw new InternalServerErrorException(
+          "THIS SHOULD NEVER HAPPEN! Error converting pipeline inputs to string", e);
+    }
+
+    // save related pipeline inputs
+    PipelineInput pipelineInput = new PipelineInput();
+    pipelineInput.setJobId(createdPipelineRun.getId());
+    pipelineInput.setInputs(pipelineInputsAsString);
+    pipelineInputsRepository.save(pipelineInput);
+
+    return createdPipelineRun;
+  }
+
+  protected PipelineRun writePipelineRunToDbThrowsDuplicateException(PipelineRun pipelineRun)
+      throws DuplicateObjectException {
+    try {
+      pipelineRunsRepository.save(pipelineRun);
+      logger.info("pipelineRun saved for jobId: {}", pipelineRun.getJobId());
+    } catch (DataIntegrityViolationException e) {
+      if (e.getCause() instanceof ConstraintViolationException c
+          && c.getConstraintName().contains("jobId_unique")) {
+        throw new DuplicateObjectException(
+            String.format("Duplicate jobId %s found", pipelineRun.getJobId()));
+      }
+      throw e;
+    }
+
+    return pipelineRun;
+  }
+
+  public PipelineRun getPipelineRun(UUID jobId, String userId) {
+    return pipelineRunsRepository.findByJobIdAndUserId(jobId, userId).orElse(null);
+  }
+
+  @Transactional
```

Comment:
why does this need to be transactional?

Reply:
🤔 not sure why that got in there. it doesn't. good catch, thanks!